Userland/cat.cpp

```diff
@@ -25,27 +25,55 @@
  */
 
 #include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
 #include <assert.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <sys/stat.h>
 #include <string.h>
 #include <unistd.h>
 
+bool detect_useless_use_of_cat(int argc)
+{
+    // I can think of somewhat valid uses of zero-argument cat
+    // from a tty to a pipe, so let's allow those.
```

Comment:
Poifect:
```c++
 if (command.argv.is_empty() && !(command.rewirings.is_empty() && command.redirections.is_empty())) {
        // Set the argv to 'cat' when some redirections exist
        // This allows `<foo' to work.
        command.argv.append("cat");
    }
```

Reply:
I would argue `<foo` shouldn't cat `foo` to stdout, it should `close(open("foo", O_RDONLY))`

Reply:
Uhhh...what's the point of that?
`<foo >bar` (as done by bash and ash) takes the stuff from foo and dumps it into bar.
I kinda like that tbh

Reply:
Pretty sure Bash doesn't do that and behaves as I described. It's not very *useful* (although it is nice to be able to create files by just saying `> foo`), but it's consistent:
* An empty command (empty argv) does nothing
* Redirections open (and close, and dup) files
* Redirections on an empty command just open files

Bash (and [POSIX shell](https://pubs.opengroup.org/onlinepubs/009695399/utilities/exec.html)) also has "exec", which is basically a way to tell it not to fork before executing the command. If you leave out the command, it doesn't actually do the `exec*()`, but it still performs redirections, which is a (somewhat) convenient way to open and close files in the shell process itself:

```
$ exec 3>/tmp/foo
$ echo hi >&3
$ exec 3>&-
```

Reply:
Hmmm...yeah, only zsh does that...weird.

what the hell is `>&-`?

Reply:
"close this fd"

Reply:
wait it keeps the filedesricptors open?
that's _weird_

Reply:
me no want that weirdness.
I can make it do nothing except resolve redirections if no argv, but you will not convince me to do _that_. redirection targets live until the command ends, otherwise it's just...hard to reason about...

Reply:
weird tho, does it leak all those fds when it forks?
how does it manage those, I wonder

Reply:
No, you don't understand. `exec` is what makes it *not fork*. Normally, you'd use it like that:

`exec some-command args 2> /dev/null`

This would make the shell execute the command as usual, except it won't fork, so basically the shell itself applies redirections and exec's in the shell process; after the command exits the process is gone.

But if you specify an empty command, it doesn't actually exec, so you keep the shell — but it performs redirections as usual, but since they happen in the Shell process, not in the fork, they persist.

Reply:
What I'm trying to say, if we take a step back, is that these parts — fork, exec/argv, file redirections/operations — are orthogonal in Bash, are controlled independently (can do file operations without exec or without fork or without either, can do exec without fork or without redirections...), and behave consistently no matter which combination you use.

Reply:
Yeah, allowing the user to mess with the descriptors of the shell itself is pretty silly.
we use 5 fds already, `3` would just overwrite some event loop thing

Reply:
No, we must absolutely support that — maybe not with that exact syntax if that's up for debate, but the general idea is that the shell is just a convenient way to *script Unix stuff* and that chiefly means forking, execing, and messing with file descriptors, which the shell is just a convenient (ugh that word again) scripting syntax for.

As for the event loop, yeah, we're going to have to make it use some high fd numbers (bash uses fds above 1024 I believe)

Reply:
I mean...if you're this persistent about it...sure, I'll let the user shoot themselves in the foot.
it feels pretty gross to me, but...whatever, sure.

_can_ we change the event loop fds?

Reply:
worst of both worlds: `$ 3>foo` applies the redirection to the shell (it's otherwise a meaningless operation anyway)